### PR TITLE
Please review my code.

### PR DESCRIPTION
**Pull request by Akshama**

### CODE REVIEW

File: 3_WordBreak.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            // consider all prefixes of current string
+            string prefix = str.substr(0, i);
+            // if prefix is found in dictionary, then recurse for suffix
+            if (find(dict.begin(), dict.end(), prefix) != dict.end() && wordBreak(dict, str.substr(i), canBeSegmented))
+                return canBeSegmented[len] = 1;
+
+        }
+    }
+    // return solution to current sub-problem
+    return canBeSegmented[len];
```
Comment (mooreaarond) [inline 3_WordBreak.cpp]: you're returning an int as a bool here, it looks like it should always be the case that the value is 0 or 1, however, it could be easy to make small changes to the code where the value would be -1 which would silently return true

File: 3_WordBreak.cpp [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+5. W - WALK THROUGH
+Explained as comments in the code
+6. I - IMPLEMENT
+*/
+#include <iostream>
+#include <string>
+#include <vector>
+#include <algorithm>
+using namespace std;
+bool wordBreak(vector<string> const &dict, string str, vector<int> &canBeSegmented)
```
Comment (mooreaarond) [inline 3_WordBreak.cpp]: In the real world, you would probably need some kind of utility function for this where the "public" signature is 
`bool wordBreak(vector<string> const &dict, string str)` and then you internally create your vector canBeSegmented and call this function here.  Since your algorithm relies on canBeSegmented having a specific size and format, you shouldn't let anyone put anything in here

File: 3_WordBreak.cpp [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+//T - TEST
+bool testWordBreak()
+{
+    vector<string> dict = { "pear", "salmon", "foot", "prints", "footprints", "leave", "you", "sun", "girl", "enjoy" ,"heart", "artist", "he", "art" };
+    string str1 = ""; //empty dict and empty string
+    string str2 = "heartist"; //string can be segmented
+    string str3 = "You"; //case sensitivity
+    vector<int> canBeSegmented1(str1.length() + 1, -1);
+    if (wordBreak({""}, "",canBeSegmented1 )!= false) //empty dict and empty string
```
Comment (mooreaarond) [inline 3_WordBreak.cpp]: Testing these together is a good case, but you should also test these conditions individually (one test with empty dictionary nonempty word and vice versa)

File: 3_WordBreak.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        cout<<"\nOne or more test cases failed";
+        return false;
+    }
+    vector<int> canBeSegmented2(str2.length() + 1, -1);
```
Comment (mooreaarond) [inline 3_WordBreak.cpp]: I think something is going on with your test cases here.  It looks like you are creating the vector based on size str2="heartist", but the test actually uses "youleavefootprints"

File: 3_WordBreak.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+#include <string>
+#include <vector>
+#include <algorithm>
+using namespace std;
+bool wordBreak(vector<string> const &dict, string str, vector<int> &canBeSegmented)
+{
+    int len = str.size();
+    if (len == 0)
+        return true;
+    //subproblem encountered for the first time
```
Comment (mooreaarond) [inline 3_WordBreak.cpp]: it might be worth having a case here for if dict.length == 0 return false, otherwise you waste the time of iterating the string.

File: 3_WordBreak.cpp (hunk 2 above)
Comment (mooreaarond) [inline 3_WordBreak.cpp]: Another note: dict should be a type that has constant lookup time.  Especially if dict is very large, this could save some major time.  I think in c++ the best option is unordered_set, but there may be other options
https://stackoverflow.com/questions/921384/java-string-array-is-there-a-size-of-method

**Review by mooreaarond — COMMENTED** [review]
I think you got it! I've left a few comments for you to look over here, but this looks pretty well done.  One thing I would ask you to focus on is my comment about unordered_set.  While you may or may not get penalized for this in an interview, choosing the right data types can have drastic performance implications in the real world.

File: 3_WordBreak.cpp (hunk 2 above)
Comment (Akshama) [reply 3_WordBreak.cpp]: Corrected.

Comment (Akshama) [reply 3_WordBreak.cpp]: Thanks for the link, I'll see if I could implement this.

File: 3_WordBreak.cpp (hunk 5 above)
Comment (Akshama) [reply 3_WordBreak.cpp]: Corrected.

File: 3_WordBreak.cpp (hunk 1 above)
Comment (Akshama) [reply 3_WordBreak.cpp]: Thanks for pointing it out! I've changed the return type of the function to int.

File: 3_WordBreak.cpp (hunk 3 above)
Comment (Akshama) [reply 3_WordBreak.cpp]: Added these test cases.

File: 3_WordBreak.cpp (hunk 4 above)
Comment (Akshama) [reply 3_WordBreak.cpp]: Corrected.
